File: frontend/src/billing.ts
/**
 * Billing facade.
 *
 * - On a native Android build (APK / dev client) → uses `expo-iap` to talk
 *   directly to Google Play Billing for the subscription products. The backend
 *   `/api/billing/purchase` endpoint then verifies the purchase token against
 *   the Google Play Developer API.
 * - On Expo Go / web preview → uses the backend mock endpoints so the
 *   paywall UI can be exercised end-to-end without a Play account.
 *
 * The choice is made at runtime via Constants.executionEnvironment.
 */
import Constants from 'expo-constants';
import { Platform } from 'react-native';

const BASE = process.env.EXPO_PUBLIC_BACKEND_URL;

export type SubscriptionStatus = {
  tier: 'free' | 'pro';
  status: 'none' | 'trialing' | 'active' | 'canceled';
  plan_id: string | null;
  is_pro: boolean;
  in_trial: boolean;
  expires_at: string | null;
  auto_renewing: boolean;
};

export type Plan = {
  id: string;
  name: string;
  price_cents: number;
  currency: string;
  period: 'month' | 'year' | 'lifetime';
  trial_days: number;
  save_pct?: number;
  limited_to?: number;
};

export type FoundersStatus = {
  remaining: number;
  total: number;
  claimed: number;
  available: boolean;
  price_cents: number;
};

const PLAN_IDS = ['premium_monthly', 'premium_annual', 'premium_lifetime'];

const isNativeBuild =
  Platform.OS === 'android' &&
  Constants.executionEnvironment !== 'storeClient' && // not Expo Go
  Constants.appOwnership !== 'expo';

// ─── Backend helpers (used in both modes for status + verify) ──────────────
async function http<T>(path: string, init?: RequestInit): Promise<T> {
  const res = await fetch(`${BASE}/api/billing${path}`, {
    headers: { 'Content-Type': 'application/json' },
    ...init,
  });
  if (!res.ok) throw new Error(`${path} failed`);
  return res.json();
}

// ─── expo-iap wrapper (loaded lazily; only used on a real device build) ────
let iap: any = null;
async function getIap() {
  if (!isNativeBuild) return null;
  if (iap) return iap;
  try {
    iap = await import('expo-iap');
    await iap.initConnection();
  } catch (e) {
    console.warn('expo-iap unavailable', e);
    iap = null;
  }
  return iap;
}

// ─── Public API ────────────────────────────────────────────────────────────
export const billing = {
  getStatus: () => http<SubscriptionStatus>('/status'),

  getPlans: () => http<Plan[]>('/plans'),

  getFoundersStatus: () => http<FoundersStatus>('/founders_remaining'),

  /**
   * Start a 7-day free trial. On native builds, this triggers Google Play's
   * subscription purchase flow (which honors the configured introductory
   * trial offer). On Expo Go / web, the backend mocks the trial start.
   */
  startTrial: async (plan_id: string): Promise<SubscriptionStatus> => {
    const i = await getIap();
    if (i) {
      // 1. Fetch the subscription details from Play.
      const subs = await i.fetchProducts({ skus: PLAN_IDS, type: 'subs' });
      const sub = subs.find((s: any) => s.productId === plan_id || s.id === plan_id);
      if (!sub) throw new Error(`SKU ${plan_id} not configured in Play Console`);
      // 2. Trigger purchase. This shows the Play UI and the user accepts trial.
      const offerToken =
        sub.subscriptionOfferDetails?.[0]?.offerToken ||
        sub.subscriptionOfferDetailsAndroid?.[0]?.offerToken;
      const purchase = await i.requestPurchase({
        request: {
          android: {
            skus: [plan_id],
            subscriptionOffers: offerToken
              ? [{ sku: plan_id, offerToken }]
              : undefined,
          },
        },
        type: 'subs',
      });
      // 3. Send purchase token to our backend for server-side verification.
      const token =
        (purchase as any)?.purchaseToken ||
        (purchase as any)?.[0]?.purchaseToken;
      return http<SubscriptionStatus>('/purchase', {
        method: 'POST',
        body: JSON.stringify({ plan_id, purchase_token: token }),
      });
    }
    // Mock path
    return http<SubscriptionStatus>('/start_trial', {
      method: 'POST',
      body: JSON.stringify({ plan_id }),
    });
  },

  purchase: async (plan_id: string): Promise<SubscriptionStatus> => {
    return billing.startTrial(plan_id); // same flow on Play; trial is part of offer
  },

  cancel: async () => {
    // Subscriptions are cancelled in Google Play, not via the app.
    if (isNativeBuild) {
      const i = await getIap();
      if (i?.deepLinkToSubscriptions) {
        await i.deepLinkToSubscriptions({ skuAndroid: 'premium_annual' });
      }
    }
    return http<SubscriptionStatus>('/cancel', { method: 'POST' });
  },

  restore: async () => {
    const i = await getIap();
    if (i) {
      const purchases = await i.getAvailablePurchases();
      // Re-validate any active purchases against backend
      for (const p of purchases) {
        if (PLAN_IDS.includes(p.productId)) {
          await http<SubscriptionStatus>('/purchase', {
            method: 'POST',
            body: JSON.stringify({
              plan_id: p.productId,
              purchase_token: p.purchaseToken,
            }),
          });
        }
      }
    }
    return http<SubscriptionStatus>('/restore', { method: 'POST' });
  },

  reset: () => http('/reset', { method: 'POST' }), // dev only
};

export const formatPrice = (cents: number, currency = 'USD'): string => {
  const amount = cents / 100;
  if (currency === 'USD') return `$${amount.toFixed(2)}`;
  return `${amount.toFixed(2)} ${currency}`;
};

export const FREE_LIMITS = {
  COACH_MESSAGES_PER_DAY: 5,
  GOALS_MAX: 2,
  STATS_HISTORY_DAYS: 7, // free sees 7 days; Pro sees 30
};

File: backend/billing.py
"""
Subscription endpoints for ScreenSense.

Real Google Play Billing verification requires `google-api-python-client`,
a service account with Android Publisher permissions, and a built APK.
In the preview environment we expose a MOCK billing endpoint that simulates
the full purchase + trial lifecycle so the paywall UI can be exercised
end-to-end. The frontend `billing` facade swaps to expo-iap automatically
when running in a development/production build.
"""
from datetime import datetime, timezone, timedelta
from typing import Optional, Dict, Any
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

from server import db  # reuse the same Motor client

billing_router = APIRouter(prefix="/api/billing", tags=["billing"])

USER_ID = "default_user"

PLANS = {
    "premium_monthly": {
        "id": "premium_monthly",
        "name": "Monthly",
        "price_cents": 499,
        "currency": "USD",
        "period": "month",
        "trial_days": 7,
    },
    "premium_annual": {
        "id": "premium_annual",
        "name": "Annual",
        "price_cents": 2999,
        "currency": "USD",
        "period": "year",
        "trial_days": 7,
        "save_pct": 50,
    },
}


class StartTrialBody(BaseModel):
    plan_id: str


class PurchaseBody(BaseModel):
    plan_id: str
    purchase_token: Optional[str] = None  # set by expo-iap in production


def _seconds(now: datetime, days: int) -> datetime:
    return now + timedelta(days=days)


@billing_router.get("/plans")
async def list_plans():
    return list(PLANS.values())


@billing_router.get("/status")
async def get_subscription_status():
    """Return current subscription state. Free if no record."""
    now = datetime.now(timezone.utc)
    doc = await db.subscriptions.find_one({"user_id": USER_ID}, {"_id": 0})
    if not doc:
        return {
            "tier": "free",
            "status": "none",
            "plan_id": None,
            "is_pro": False,
            "in_trial": False,
            "expires_at": None,
            "auto_renewing": False,
        }
    expires_at = doc.get("expires_at")
    if isinstance(expires_at, str):
        expires_at = datetime.fromisoformat(expires_at)
    # MongoDB sometimes returns offset-naive datetimes — coerce to UTC.
    if isinstance(expires_at, datetime) and expires_at.tzinfo is None:
        expires_at = expires_at.replace(tzinfo=timezone.utc)
    in_trial = doc.get("in_trial", False) and expires_at and expires_at > now
    active = expires_at and expires_at > now and doc.get("status") in ("trialing", "active")
    return {
        "tier": "pro" if active else "free",
        "status": doc.get("status", "none"),
        "plan_id": doc.get("plan_id"),
        "is_pro": bool(active),
        "in_trial": bool(in_trial),
        "expires_at": expires_at.isoformat() if expires_at else None,
        "auto_renewing": doc.get("auto_renewing", True),
    }


@billing_router.post("/start_trial")
async def start_trial(body: StartTrialBody):
    """Begin a 7-day free trial (no card required in preview)."""
    if body.plan_id not in PLANS:
        raise HTTPException(400, "Invalid plan")
    plan = PLANS[body.plan_id]
    now = datetime.now(timezone.utc)

    existing = await db.subscriptions.find_one({"user_id": USER_ID}, {"_id": 0})
    if existing and existing.get("status") in ("active", "trialing"):
        raise HTTPException(409, "Already subscribed")
    if existing and existing.get("trial_used"):
        raise HTTPException(409, "Trial already used. Upgrade directly.")

    record = {
        "user_id": USER_ID,
        "plan_id": body.plan_id,
        "status": "trialing",
        "in_trial": True,
        "trial_used": True,
        "trial_started_at": now.isoformat(),
        "expires_at": _seconds(now, plan["trial_days"]),
        "auto_renewing": True,
        "purchase_token": None,
        "updated_at": now.isoformat(),
    }
    await db.subscriptions.update_one(
        {"user_id": USER_ID}, {"$set": record}, upsert=True)
    return await get_subscription_status()


@billing_router.post("/purchase")
async def record_purchase(body: PurchaseBody):
    """
    Record a purchase. In production the `purchase_token` would be verified
    via the Google Play Developer API; the preview accepts any token.
    """
    if body.plan_id not in PLANS:
        raise HTTPException(400, "Invalid plan")
    plan = PLANS[body.plan_id]
    now = datetime.now(timezone.utc)
    days = 30 if plan["period"] == "month" else 365
    record = {
        "user_id": USER_ID,
        "plan_id": body.plan_id,
        "status": "active",
        "in_trial": False,
        "trial_used": True,
        "expires_at": _seconds(now, days),
        "auto_renewing": True,
        "purchase_token": body.purchase_token or f"mock-{plan['id']}-{int(now.timestamp())}",
        "updated_at": now.isoformat(),
    }
    await db.subscriptions.update_one(
        {"user_id": USER_ID}, {"$set": record}, upsert=True)
    return await get_subscription_status()


@billing_router.post("/cancel")
async def cancel_subscription():
    """Cancel auto-renew. User keeps access until expires_at."""
    now = datetime.now(timezone.utc)
    await db.subscriptions.update_one(
        {"user_id": USER_ID},
        {"$set": {"auto_renewing": False, "updated_at": now.isoformat()}}
    )
    return await get_subscription_status()


@billing_router.post("/restore")
async def restore_purchase():
    """Restore - in mock mode just returns current status."""
    return await get_subscription_status()


@billing_router.post("/reset")
async def reset_subscription():
    """DEV ONLY: clear subscription state so trial can be tested again."""
    await db.subscriptions.delete_many({"user_id": USER_ID})
    return {"reset": True}
